# 간단한 몇 가지

## → array, list, dictionary로 바꾸기

In [1]:
import pydataset
mpg = pydataset.data('mpg')

df = mpg.iloc[:5, 2:5]
df

displ  year  cyl
1    1.8  1999    4
2    1.8  1999    4
3    2.0  2008    4
4    2.0  2008    4
5    2.8  1999    6

In [2]:
#1. df를 array로 바꾸기
df.values

array([[1.800e+00, 1.999e+03, 4.000e+00],
       [1.800e+00, 1.999e+03, 4.000e+00],
       [2.000e+00, 2.008e+03, 4.000e+00],
       [2.000e+00, 2.008e+03, 4.000e+00],
       [2.800e+00, 1.999e+03, 6.000e+00]])

In [3]:
#2. df를 딕셔너리로 바꾸기
df.to_dict()

{'displ': {1: 1.8, 2: 1.8, 3: 2.0, 4: 2.0, 5: 2.8},
 'year': {1: 1999, 2: 1999, 3: 2008, 4: 2008, 5: 1999},
 'cyl': {1: 4, 2: 4, 3: 4, 4: 4, 5: 6}}

In [4]:
#3. df를 리스트로 바꾸기
df.values.tolist() #리스트로 바꾸려면 데이터프레임일때는 X

[[1.8, 1999.0, 4.0],
 [1.8, 1999.0, 4.0],
 [2.0, 2008.0, 4.0],
 [2.0, 2008.0, 4.0],
 [2.8, 1999.0, 6.0]]

## → 리스트를 가로 방향으로 한 줄 세우기

In [5]:
#df를 하나의 리스트로 만들기(가로로 줄 세우기)
li = []
for i in df.values.tolist():    #정석 방식
    for j in i:
        li.append(j)
li

[1.8,
 1999.0,
 4.0,
 1.8,
 1999.0,
 4.0,
 2.0,
 2008.0,
 4.0,
 2.0,
 2008.0,
 4.0,
 2.8,
 1999.0,
 6.0]

In [6]:
df.values.reshape(-1).tolist()   #같은방법

[1.8,
 1999.0,
 4.0,
 1.8,
 1999.0,
 4.0,
 2.0,
 2008.0,
 4.0,
 2.0,
 2008.0,
 4.0,
 2.8,
 1999.0,
 6.0]

In [7]:
df.values.flatten().tolist() #flatten() '평평하게 편다'

[1.8,
 1999.0,
 4.0,
 1.8,
 1999.0,
 4.0,
 2.0,
 2008.0,
 4.0,
 2.0,
 2008.0,
 4.0,
 2.8,
 1999.0,
 6.0]

# 사이킷런

In [8]:
import sklearn

## □ 붓꽃 품종 예측하기 

### 1. 데이터 세트 분리
: 데이터를 학습 데이터와 테스트 데이터로 분리합니다.

In [35]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split  #분리

iris = load_iris()
iris

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

In [13]:
type(iris) 
#Bunch라고 출력되지만 사실은 딕셔너리와 같은것

sklearn.utils.Bunch

In [ ]:
iris.keys()
iris.values()
iris.items() #key와 values를 동시에 보여준다.

In [15]:
iris_data = iris['data']
iris_data  #딕셔너리에서 특정 값을 찾는 것과 같은 방식

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

In [16]:
iris_label = iris['target']
iris_label

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [18]:
X_train, X_test, y_train, y_test = train_test_split(iris_data, iris_label)
# iris_data -> X값   iris_label -> y값

In [21]:
X_train.shape

(112, 4)

데이터의 갯수는 112개라는 이야기.

칼럼.특질.피쳐는 4개이다.

#### → 주요 Key의 의미

```
- data : X(피처)의 데이터 세트

- target : 최종적으로 구하려는 결괏값
           (분류 시에는 레이블 값, 회귀 시에는 숫자 결괏값 데이터 세트로 나타난다.)

- target_names : 개별 y(레이블)의 이름, 분류하고자 하는 대상

- feature_names : 피처의 이름

- DESCR : describe의 약자로 데이터에 대한 설명
```

#### → train_test_split() Parameter

In [ ]:
from sklearn.model_selection import train_test_split
train_test_split(arrays, test_size, train_size, random_state, shuffle, stratify)

```
- arrays : 분할시킬 데이터를 입력

- test_size : 테스트 데이터셋의 비율(float)이나 갯수(int) (default = 0.25)

- train_size : 학습 데이터셋의 비율(float)이나 갯수(int) (default = test_size의 나머지)

- random_state : 데이터 분할시 셔플이 이루어지는데 이를 위한 시드값 (int나 RandomState로 입력)

- shuffle : 셔플여부설정 (default = True)

- stratify : 지정한 Data의 비율을 유지한다. 
             예를 들어, Label Set인 Y가 25%의 0과 75%의 1로 이루어진 Binary Set일 때, 
             stratify=Y로 설정하면 나누어진 데이터셋들도 0과 1을 각각 25%, 75%로 유지한 채 분할된다.
```

### 2. 모델 학습
: 학습 데이터를 기반으로 ML 알고리즘을 적용해 모델을 학습시킵니다.

In [36]:
dt_clf = DecisionTreeClassifier()
# DecisionTreeClassifier 객체 생성(아직 아무런 기능 없다.)

In [20]:
dt_clf.fit(X_train, y_train)
#fitting 적합한다. 학습한다. 훈련한다.

DecisionTreeClassifier()

In [32]:
X_test.shape  #38개가 있다.

(38, 4)

### 3. 예측 수행
: 학습된 ML 모델을 이용해 테스트 데이터의 분류(즉, 붓꽃 종류)를 예측합니다.

In [23]:
y_pred = dt_clf.predict(X_test)
y_pred  #예측

array([2, 1, 0, 2, 0, 0, 2, 2, 1, 1, 0, 2, 2, 1, 0, 2, 0, 1, 2, 0, 2, 1,
       1, 1, 2, 1, 1, 0, 1, 0, 2, 1, 0, 0, 2, 0, 0, 1])

In [24]:
y_test  #정답

array([1, 1, 0, 2, 0, 0, 2, 2, 1, 1, 0, 2, 2, 1, 0, 2, 0, 1, 2, 0, 2, 2,
       1, 1, 2, 1, 1, 0, 1, 0, 2, 1, 0, 0, 2, 0, 0, 1])

In [33]:
y_pred == y_test  # 둘 간 얼마나 일치한가?

array([False,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True, False,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True])

In [34]:
sum(y_pred == y_test) #일치 갯수는 36개, 처음에는 38개였다.

36

In [28]:
iris.target_names

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

위에 y_test 출력값을 보면 1,0 이런식으로 되어있는데

이 숫자가 의미하는 바는 위에 어레이의 인덱스를 표현한 것이다.

### 4. 평가
: 이렇게 예측된 결괏값과 테스트 데이터의 실제 결괏값을 비교해 ML 모델 성능을 평가한다.

In [30]:
from sklearn.metrics import accuracy_score  #정확도를 보여준다.

In [31]:
accuracy_score(y_test, y_pred)

0.9473684210526315

# 데이터 전처리
```
< 이유 >

- 결손값, 즉 NaN, Null 값은 허용되지 않습니다. (이러한 경우 pandas로 처리)
- 문자열 값을 입력값으로 허용하지 않습니다. (그래서 0,1 이런식으로 표현함.)

```

## □ 레이블 인코딩
- LabelEncoder를 객체로 생성 후 fit()과 transform()을 호출해 인코딩

In [39]:
from sklearn.preprocessing import LabelEncoder

encoder = LabelEncoder()

In [40]:
labels = encoder.fit_transform(['TV', '냉장고', '전자레인지'])
labels

array([0, 1, 2], dtype=int64)

## □ 원-핫 인코딩
One-Hot Encoding / '한 개만 불이 들어와요'